File: fpinjava-parent/fpinjava-recursion-exercises/src/main/java/com/fpinjava/recursion/exercise04_06/ComposeAll.java
package com.fpinjava.recursion.exercise04_06;

import java.util.List;

import com.fpinjava.common.CollectionUtilities;
import com.fpinjava.common.Function;

import static com.fpinjava.common.CollectionUtilities.foldRight;
import static com.fpinjava.common.Function.identity;


public class ComposeAll {

	static <T> Function<T, T> composeAll(List<Function<T, T>> list) {
		return foldRight(list, identity(), x -> x::compose);
	}
}

File: fpinjava-parent/fpinjava-recursion-exercises/src/main/kotlin/com/fpinkotlin/recursion/exercise04_06/ComposeAll.kt
package com.fpinkotlin.recursion.exercise04_06

import com.fpinkotlin.common.compose
import com.fpinkotlin.recursion.exercise04_05.foldRight

fun <T> composeAll(list: List<(T) -> (T)>): (T) -> (T) =
		foldRight(list, { x -> x }) { x -> { f: (T) -> T -> x.compose(f) } }
